blockchain/chain/src/store/chain_store.rs

```diff
@@ -273,6 +278,52 @@ where
         out += &value;
         Ok(out)
     }
+
+    pub fn get_randomness_block(
+        &self,
+        blocks: &TipsetKeys,
+        pers: DomainSeparationTag,
+        round: ChainEpoch,
+        entropy: &[u8],
+    ) -> Result<Vec<u8>, Error> {
+        get_randomness(self.db.as_ref(), blocks, pers, round, entropy)
+            .map_err(|e| Error::Other(e.to_string()))
+    }
+}
+
+/// Gets 32 bytes of randomness for ChainRand paramaterized by the DomainSeparationTag, ChainEpoch, Entropy
+pub fn get_randomness<DB: BlockStore>(
+    db: &DB,
+    blocks: &TipsetKeys,
+    pers: DomainSeparationTag,
+    round: ChainEpoch,
+    entropy: &[u8],
+) -> Result<Vec<u8>, Box<dyn std::error::Error>> {
+    let mut blks = blocks.clone();
+    loop {
+        let nts = tipset_from_keys(db, &blks)?;
+        let mtb = nts.min_ticket_block();
+        if nts.epoch() <= round || mtb.epoch() == 0 {
+            return draw_randomness(mtb.ticket().vrfproof.bytes(), pers, round, entropy);
+        }
+        blks = mtb.parents().clone();
+    }
+}
+
+/// Computes a pseudorandom 32 byte Vec
+fn draw_randomness(
+    rbase: &[u8],
+    pers: DomainSeparationTag,
+    round: ChainEpoch,
+    entropy: &[u8],
+) -> Result<Vec<u8>, Box<dyn std::error::Error>> {
```

Comment:
Can you just make the return a 32 byte array to remove the need for allocations?